inluxclient/client.go

```diff
@@ -71,3 +91,83 @@ func New(params Params) (*Client, error) {
 	}
 	return c, nil
 }
+
+// makeAPICall issues an HTTP request to InfluxDB server API url according to parameters.
+// Additionally, sets Authorization header and User-Agent.
+// It returns http.response or Error
+func (c *Client) makeAPICall(ctx context.Context, params httpParams) (*http.Response, *ServerError) {
+	// copy URL
+	urlObj := *params.endpointURL
+	urlObj.RawQuery = params.queryParams.Encode()
+
+	fullURL := urlObj.String()
+
+	req, err := http.NewRequestWithContext(ctx, params.httpMethod, fullURL, params.body)
+	if err != nil {
+		return nil, NewServerError(fmt.Sprintf("error calling %s: %v", fullURL, err))
+	}
+	for k, v := range params.headers {
+		req.Header.Set(k, v)
+	}
+	req.Header.Set("User-Agent", userAgent)
+	if c.authorization != "" {
+		req.Header.Add("Authorization", c.authorization)
+	}
+
+	resp, err := c.params.HTTPClient.Do(req)
+	if err != nil {
+		return nil, NewServerError(fmt.Sprintf("error calling %s: %v", fullURL, err))
+	}
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, c.resolveHTTPError(resp)
+	}
+
+	return resp, nil
+}
+
+// resolveHTTPError parses server error response and returns error with human readable message
+func (c *Client) resolveHTTPError(r *http.Response) *ServerError {
+	// successful status code range
+	if r.StatusCode >= 200 && r.StatusCode < 300 {
+		return nil
+	}
+
+	httpError := &ServerError{
```

Comment:
I think we could probably try both error kinds here with one decode, which could simplify the logic quite a bit:

```
var httpError struct {
     ServerError
     // This is used by the InfluxDB v1 responses.
     Error string `json:"error"`
}
httpError.StatusCode = r.StatusCode
... decode into httpError
if httpError.Code == "" && httpError.Message == "" {
    httpError.Message = httpError.Error
}

```

Reply:
Ok